[Fix rubocop#9328] Honour shareable_constant_value magic comment

thearjunmdas opened the pull request:
Ruby 3.0 supports [shareable_constant_value](https://docs.ruby-lang.org/en/3.0.0/doc/syntax/comments_rdoc.html#label-shareable_constant_value+Directive)  magic comment. This PR updates Style/MutableConstant to honour shareable_constant_value magic comment.

Fixes [#9328](https://github.com/rubocop/rubocop/issues/9328)

Took PR [9410](https://github.com/rubocop/rubocop/pull/9410) as reference. Tried to add @caalberts as co-author in changelog, but changelog spec fails.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

thearjunmdas commented:
@marcandre  - Could you please take a look when free?

marcandre commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -157,6 +157,123 @@
+      context 'when using shareable_constant_value: literal' do
+        let(:prefix) { '# shareable_constant_value: literal' }
+
+        it_behaves_like 'immutable objects', '[1, 2, 3]'
+        it_behaves_like 'immutable objects', '%w(a b c)'
+        it_behaves_like 'immutable objects', '{ a: 1, b: 2 }'
+        it_behaves_like 'immutable objects', "'str'"
+        it_behaves_like 'immutable objects', '"top#{1 + 2}"'
+        it_behaves_like 'immutable objects', '1'
+        it_behaves_like 'immutable objects', '1.5'
+        it_behaves_like 'immutable objects', ':sym'
+        it_behaves_like 'immutable objects', 'FOO + BAR'
+        it_behaves_like 'immutable objects', 'FOO - BAR'
+        it_behaves_like 'immutable objects', "'foo' + 'bar'"
+        it_behaves_like 'immutable objects', "ENV['foo']"
+        it_behaves_like 'immutable objects', "::ENV['foo']"
```
Could these be grouped together and reused instead of copied?

thearjunmdas replied:
Thanks for the comment. Refactored.

marcandre commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -157,6 +157,123 @@
+        RUBY
+      end
+
+      it 'raises offense only for shareable_constant_value as none when set in the order of: literal, none and experimental_everything' do
+        expect_offense(<<~RUBY)
+          # shareable_constant_value: literal
+          X = [1, 2, 3]
+          # shareable_constant_value: none
+          Y = [4, 5, 6]
+              ^^^^^^^^^ Freeze mutable objects assigned to constants.
+          # shareable_constant_value: experimental_everything
+          Z = [7, 8, 9]
+        RUBY
```
This test seems sufficient to me, all the others look like variations.

thearjunmdas replied:
Agreed. Refactored.

marcandre commented on lib/rubocop/cop/mixin/shareable_constant_value.rb:
```diff
@@ -0,0 +1,37 @@
+    module ShareableConstantValue
+      module_function
+
+      def recent_shareable_value?(node)
```
This methods has bad performance (`O(n)`) so overall this cop could be `O(n^2)`.
Ideally fix the performance of this method, or else make sure that it is called *last*, i.e. only when it really looks like there's a `.freeze` missing.

thearjunmdas replied:
For now called this as last line to make the best effort on performance.

I believe the only way to make this method performant is to memoize, rather store the last instance shareable_constant_value was provided. Is there any place in the codebase where such memoization is already being done? I could refer that raise a future PR to improve the performance of this method

marcandre replied:
You can use normal memoization (`@cache ||= ...`). To be clean, you can clear the cache in `def on_new_investigation`, but it's not necessary unless your cop `support_multiple_source?`.

marcandre commented on lib/rubocop/cop/style/mutable_constant.rb:
```diff
@@ -132,6 +134,13 @@ def frozen_string_literal?(node)
 
+        def shareable_constant_value?(node)
+          return false if target_ruby_version < 3.0
+          return false if node.nil?
```
Is there a valid reason for this method to ever be called with `node == nil`?

thearjunmdas replied:
All the scenarios like the following https://github.com/rubocop/rubocop/blob/master/spec/rubocop/cop/style/mutable_constant_spec.rb#L28 passes a nil node to this

marcandre replied:
Oh, I see. That's a mistake, fixed in #9931

marcandre reviewed (commented):
Tests that could be added

1)
```ruby
# inline comments:
X = "x" # shareable_constant_value: literal
```

2)
```ruby
# shareable_constant_value: literal
X = Set[1, 2]
```

3)
```ruby
# blabla shareable_constant_value: literal blabla
```

4) variations with hyphen

marcandre commented:
Good PR @thearjunmdas ! 👍
I left a few comments for you

thearjunmdas commented:
> Good PR @thearjunmdas ! 👍
> I left a few comments for you

Thanks a lot for the quick review. Will address the comments and get back.

thearjunmdas commented:
https://github.com/rubocop/rubocop/pull/9929#pullrequestreview-705204492 - Added all the test cases mentioned here.
Addressed all other review comments @marcandre. I am new to RSpec, please check if the way I have refactored is okay.

Please take a look and feel free to add more comments if we could make it better.

marcandre commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -43,6 +43,45 @@
     end
   end
 
+  shared_examples 'freezing all constants' do |o|
```
Using `shared_examples` is the right way to go 👍
OTOH, the label or grouping could be improved. "not freezing all constants" is not clear. Something like "enforces mutable literals are frozen" seems more precise, for example.

thearjunmdas replied:
I was trying to make it a readable sentence with **it_behaves_like**. Please do check if this makes sense
_it_behaves_like 'literals that are frozen'_
Also mentioning literals since i would like to include both mutable and immutable literals.

marcandre replied:
There are three categories of values assigned to constants:
1) immutable literals (e.g. `42`, `:symbol`)
2) mutable builtin literals (e.g. `[1, 2, 3]`)
3) general expressions (e.g. `FOO + BAR`, `Set[1, 2, 3]`)

The cop should raise for:
1) never
2) for `none`
3) for `literal` or `none`

I'd like the cop and the specs to reflect this.

Maybe something like "This cop behaves_like:
1) "it_considers_immutable_literals_frozen"
2) "it_considers_builtin_literals_frozen"
3) "it_considers_expressions_in_constants_frozen"

thearjunmdas replied:
Thanks for the detailed comments. I guessed I rushed through the tests a bit, will take a deeper look into this and come back.

thearjunmdas replied:
```ruby
# When EnforcedStyle: literal
# +-------------------+--------------------------------+----------------+
# |                   | Cop offense raised for shareable_constant_value |
# |   Values          |------+---------+---------------+----------------+
# |                   | none | literal | experimental_ | experimental_  |
# |                   |      |         | everything    | copy           |
# +-------------------+------+---------+---------------+----------------+
# |Immutable literals |false |false    |false          |false           |
# |Examples:          |      |         |               |                |
# |42, :symbol        |      |         |               |                |
# +-------------------+------+---------+---------------+----------------+
# |Mutable literals   |true  |false    |false          |false           |
# |Examples:          |      |         |               |                |
# |[1, 2, 3, 4]       |      |         |               |                |
# +-------------------+------+---------+---------------+----------------+
# |General expressions|true  |true     |false          |false           |
# |Examples:          |      |         |               |                |
# |FOO + BAR          |      |         |               |                |
# +-------------------+------+---------+---------------+----------------+
# |Frozen expressions |      |         |               |                |
# |Examples:          |false | false   | false         | false          |
# |(anything).freeze  |      |         |               |                |
# +-------------------+------+---------+---------------+----------------+

# When EnforcedStyle: Strict, Only the following row changes
# +-------------------+------+---------+---------------+----------------+
# |General Expressions|true  |false    |false          |false           |
# |                   |      |         |               |                |
# |Examples: FOO + BAR|      |         |               |                |
# +-------------------+------+---------+---------------+----------------+
```

I am planning to add this comment - Did i summarise correctly? I will go ahead once you are okay with above @marcandre.

marcandre replied:
My observation was more about naming your shared behaviours. If it's not clear to you, we can go ahead as is and eventually refactor.

thearjunmdas replied:
I tried refactoring by breaking it upto smaller chunks - Dint work out.

For now changed the title a bit, will try to refactor this over time.

marcandre commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -157,6 +196,32 @@
+      end
+
+      it 'raises offense when shareable_constant_value is specified as an inline comment' do
+        expect_offense(<<~RUBY)
+          X = [1, 2, 3] # shareable_constant_value: literal
+              ^^^^^^^^^ Freeze mutable objects assigned to constants.
+        RUBY
```
Did you verify what Ruby does for inline comments? How does it affect subsequent constants? I'm not sure what it does, and [the doc](https://docs.ruby-lang.org/en/3.0.0/doc/syntax/comments_rdoc.html#label-Scope) doesn't say. Do you feel like opening a bug report or proposing a PR for Ruby about this, or should I do it?
Also, the comments are actually scoped; current code does not take that into account. This limitation should be documented.

thearjunmdas replied:
yes @marcandre - verified in my local. The above test is  the observed behaviour.
``` ruby
A = %w[anything here is fine even non-literals] # shareable_constant_value: literal 
puts A.frozen? # prints false
```

I can open a pr in ruby for the docs. And yes, that was a good catch, i first went to the docs after your review comment but not help there.

marcandre replied:
Could you test how it affects constants that follow please (both in Ruby and in this PR)?

thearjunmdas replied:
Verified it in ruby 3.0 and updated the test in PR. 
**inline comment is not recognised.**

thearjunmdas replied:
In addition, I was trying to raise PR for this in ruby documentation - any tips on which repo and where I should go about this? https://github.com/ruby/ruby/blob/master/doc/contributing.rdoc is not very clear where to contribute for docs. https://github.com/ruby/docs.ruby-lang.org - this led me no where as well.

marcandre replied:
I guess it's ok; magic comments need always be alone on a line

thearjunmdas commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -43,6 +43,43 @@
+    it_behaves_like 'immutable objects', "ENV['foo']"
+    it_behaves_like 'immutable objects', "::ENV['foo']"
+  end
+
```
@marcandre - Made a change here. Was reviewing code and found that 
```ruby
# shareable_constant_value: literal
C = Set[2, 3] # Raises this error "cannot assign unshareable object to C (Ractor::IsolationError)"
```

I think we would need to write a separate cop for shareable_constant_value to handle the real purpose of shareable constant - ractor safe objects.

marcandre replied:
Yes, `Set[]` is not (yet) considered like a "builtin" literal, so it is normal it behaves differently from Array/Hash/...

marcandre commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -43,6 +43,43 @@
+    it_behaves_like 'immutable objects', '1'
+    it_behaves_like 'immutable objects', '1.5'
+    it_behaves_like 'immutable objects', ':sym'
+    it_behaves_like 'immutable objects', 'FOO + BAR'
+    it_behaves_like 'immutable objects', 'FOO - BAR'
+    it_behaves_like 'immutable objects', "'foo' + 'bar'"
+    it_behaves_like 'immutable objects', "ENV['foo']"
+    it_behaves_like 'immutable objects', "::ENV['foo']"
```
I don't understand these. `X = FOO + BAR` should raise an offense (unless `experimental_*`), as they won't be frozen in general, e.g.:

```ruby
FOO = [1, 2, 3].freeze
BAR = [1, 2, 3].freeze
X = FOO + BAR
X.frozen? # => false
```

thearjunmdas replied:
Yes @marcandre - I guess you are right,  tt should raise an offense. Looks like, in the current behaviour it does not. Can you check the following tests:
https://github.com/rubocop/rubocop/blob/master/spec/rubocop/cop/style/mutable_constant_spec.rb#L58

thearjunmdas replied:
Found the issue - 
The mutable constant cop relies on [unless value.mutable_literal?](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/style/mutable_constant.rb#L122) whose definition is [here](https://github.com/rubocop/rubocop-ast/blob/master/lib/rubocop/ast/node.rb#L361). This is actually not used properly. 

@marcandre  - I think we should create another bug for this and solve it for this cop.

marcandre replied:
Oh, right. It's ok to not change the current behaviour then, and yes, it could be discussed elsewhere 👍

thearjunmdas commented:
Have added some comments, please review them - then I can proceed with the changes.

marcandre commented on lib/rubocop/cop/mixin/shareable_constant_value.rb:
```diff
@@ -0,0 +1,37 @@
+module RuboCop
+  module Cop
+    # Common functionality for dealing with Shareable Constant Value
+    module ShareableConstantValue
```
Could you please:
1) Add a comment about performance being bad
2) Move this the the cop, in a private constant? I would rather not expose this code as it is.

thearjunmdas replied:
Moved this to the cop.

thearjunmdas commented on lib/rubocop/cop/style/mutable_constant.rb:
```diff
@@ -53,6 +53,43 @@ module Style
+            %w[literal experimental_everything experimental_copy].include? value
+          end
+        end
+        private_constant :ShareableConstantValue
```
@marcandre  - Please check this moved the non-performant module internal to the class and marked as private_constant.

marcandre replied:
👍

thearjunmdas commented:
Please do take a look, @marcandre

thearjunmdas commented:
ping @marcandre

thearjunmdas commented:
@koic @marcandre  - can you folks please take a look? lets try to merge this before it becomes stale?

dvandersluis commented on changelog/change_recognize_shareable_constant_value_magic.md:
```diff
@@ -0,0 +1 @@
+* [#9328](https://github.com/rubocop/rubocop/issues/9328): Recognize shareable_constant_value magic comment. ([@thearjunmdas][])
```
If you wanted to add a second person here, you could do something like this:

```suggestion
* [#9328](https://github.com/rubocop/rubocop/issues/9328): Recognize shareable_constant_value magic comment. ([@thearjunmdas][], [@caalberts][])
```

thearjunmdas replied:
@dvandersluis thanks for the tip. Worked perfectly.

bbatsov commented:
@thearjunmdas Please, also document the support for this magic comment in the cop's description, so people would know what to expect.

bbatsov commented:
You'll also have to rebase on top of `master`.

marcandre commented:
Oh, so sorry I missed the pings 😅

thearjunmdas commented:
Thanks for reviving the PR @marcandre and @bbatsov. I will update  on the comments.

P.S. Cleaning the commit history - squashing my commits, rebasing - please ignore intermittent commit history change. Will clean it up.

thearjunmdas commented:
@bbatsov @marcandre rebased on top of master, added documentation. Please let me know, if anything else needs to be done.

bbatsov commented on lib/rubocop/cop/style/mutable_constant.rb:
```diff
@@ -14,6 +14,11 @@ module Style
       # positives. Luckily, there is no harm in freezing an already
       # frozen object.
       #
+      # From Ruby 3.0, this cop honours the magic comment
```
Please, also add a couple of code snippets with `shareable_constant_value` in the cop examples.

bbatsov commented on spec/rubocop/cop/style/mutable_constant_spec.rb:
```diff
@@ -157,6 +194,34 @@
+
+      it 'raises offense when shareable_constant_value is specified as an inline comment' do
+        expect_offense(<<~RUBY)
+          X = [1, 2, 3] # shareable_constant_value: literal
```
At some point we'll need a lint cop to check for this. :D

thearjunmdas replied:
Makes, sense. Since it's an experimental feature, maybe we need not do it right away.

thearjunmdas commented on lib/rubocop/cop/style/mutable_constant.rb:
```diff
@@ -52,7 +57,59 @@ module Style
+      # NOTE: This special directive helps to create constants
+      # that hold only immutable objects, or Ractor-shareable
+      # constants. - ruby docs
+      #
```
@bbatsov - Added some examples. Also added an excerpt from ruby docs so that shareable_constant_value doesn't turn out to be an alternative to **.freeze**

thearjunmdas commented:
@bbatsov  - please take  a look.

bbatsov commented:
Looks good. Thanks!
